ntoskrnl/ke/i386/mproc.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Architecture specific source file to hold multiprocessor functions
+ * COPYRIGHT:   Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ *              Copyright 2021 Victor Perevertkin <victor.perevertkin@reactos.org>
+ */
+
+/* INCLUDES *****************************************************************/
+
+#include <ntoskrnl.h>
+#define NDEBUG
+#include <debug.h>
+
+typedef struct _APINFO
+{
+    DECLSPEC_ALIGN(PAGE_SIZE) KGDTENTRY Gdt[128];
+    DECLSPEC_ALIGN(PAGE_SIZE) KIDTENTRY Idt[256];
+    DECLSPEC_ALIGN(16) UINT8 NMIStackData[DOUBLE_FAULT_STACK_SIZE];
+    KIPCR Pcr;
+    ETHREAD Thread;
+    KTSS Tss;
+    KTSS TssDoubleFault;
+    KTSS TssNMI;
+} APINFO, *PAPINFO;
+
+
+typedef struct _AP_SETUP_STACK
+{
+    PVOID ReturnAddr;
+    PVOID KxLoaderBlock;
+} AP_SETUP_STACK, *PAP_SETUP_STACK;
+
+/* FUNCTIONS *****************************************************************/
+
+CODE_SEG("INIT")
+VOID
+NTAPI
+KeStartAllProcessors(VOID)
+{
+    PVOID KernelStack, DPCStack;
+    SIZE_T ProcessorCount = 0;
+    PAPINFO APInfo;
+
+    while (TRUE)
+    {
+        ProcessorCount++;
+        KernelStack = NULL;
+        DPCStack = NULL;
+
+        // Allocate structures for a new CPU.
+        APInfo = ExAllocatePoolZero(NonPagedPool, sizeof(APINFO), '  eK');
+        ASSERT(ALIGN_DOWN_POINTER_BY(APInfo, PAGE_SIZE) == APInfo);
```

Comment:
This does not look like a robust allocation.

Reply:
certainly needs a NULL check, indeed.

Reply:
The NULL check is right below. And NULL is also page aligned 😉

Reply:
Wait what is incorrect then? 

Is it not robust because of my use of ALIGN_DOWN_POINTER_BY?

Reply:
It's just a bit weird to check a pointer for alignment before checking it against NULL.